ARCs/arc-0023.md

```diff
@@ -0,0 +1,244 @@
+        { "type": "uint64", "name": "a", "desc": "The first factor to multiply" },
+        { "type": "uint64", "name": "b", "desc": "The second factor to multiply" }
+      ],
+      "returns": { "type": "uint64", "desc": "The product of a and b" }
+    }
+  ]
+}'''
+```
+
+and the associated TEAL script:
+```python
+from pyteal import *  # Python 3.8.3, pyteal==0.10.1
+
+calc_teal = compileTeal(
+            Cond(
+                [
+                    Txn.application_args.length() == Int(0),
+                    Cond(
+                        [Txn.application_id() == Int(0), Approve()],
+                        [Txn.on_completion() == OnComplete.OptIn, Approve()],
+                        [Txn.on_completion() == OnComplete.CloseOut, Approve()],
+                        [Txn.on_completion() == OnComplete.DeleteApplication, Reject()],
+                        [Txn.on_completion() == OnComplete.UpdateApplication, Reject()],
+                    )
+                ],
+                [
+                    Txn.on_completion() == OnComplete.NoOp,
+                    Cond(
+                        [
+                            Txn.application_args[0] == Bytes('base16', 'fe6bdf69'),  # add(uint64,uint64)uint64
+                            Seq([
+                                Log(
+                                    Concat(
+                                        Bytes('base16', '151f7c75'),
+                                        Itob(Add(Btoi(Txn.application_args[1]), Btoi(Txn.application_args[2])))
+                                    )
+                                ),
+                                Approve()
+                            ])
+                        ],
+                        [
+                            Txn.application_args[0] == Bytes('base16', '766083a7'),  # multiply(uint64,uint64)uint64
+                            Seq([
+                                Log(
+                                    Concat(
+                                        Bytes('base16', '151f7c75'),
+                                        Itob(Mul(Btoi(Txn.application_args[1]), Btoi(Txn.application_args[2])))
+                                    )
+                                ),
+                                Approve()
+                            ])
+                        ],
+
+                    )
+                ]
+            ), mode=Mode.Application, version=5
+        )
+```
+
+With the following python script we can generate the hash and append it to the compiled application, according to arc-23:
+```python
+import hashlib
+import msgpack  # Python 3.8.3, msgpack==1.0.2
+
+def sha512_256(data: bytes):
+    return hashlib.new(name='sha512_256', data=data).digest()
+
+def hash_contract_descriptor(contract_descriptor: bytes):
+    return sha512_256(b'arc0023' + contract_descriptor)
+
+def encode_arc23_object(contract_descriptor: bytes):
+    return msgpack.packb({'arc23': hash_contract_descriptor(contract_descriptor)})
```

Comment:
The message you're adding to the end of the program is not just a hash, it's the msgpack encoding of the object `{"arc23": <32-byte-hash>}`? Is the assumption that other ARCs might want to add similar things, in which case more keys on that object would be present, such as `{"arc23": <32-byte-hash>, "arc999": <32-byte-hash>}`? If that's the reason, here are some downsides:
* This is much less space efficient, since every "standard" that a contract wants to support takes _at least_ 32 bytes each.
* Since this object gets encoded as a variable-length byte string, it will be harder for clients to properly decode it. You can't just look at the last `N` bytes, you need to actually decode every opcode in the program, look at the last one, and be able to decode msgpack. That's a lot of requirements for every client to follow.

Consider this slightly different approach:
* Instead of each "standard" being applied independently to a program, create a unified JSON object that contains all the information from each standard.
* This object could have keys for each standard, and the values would be the full metadata for that standard. For example:
```json
{
  "arc23": { "name": "..." },
  "arc999": { "field_specific_to_arc999": "value..." }
}
```
* To apply this information to the contract, hash the entire object and place only the 32-byte hash at the end of the program using `bytecblock` or `pushbytes`. This way, clients need only look at the last 32-bytes of a program to prove that it implements a set of standards. No msgpack decoding or program op parsing is required.

Reply:
Another option, which impacts the earlier discussion about using `bytecblock` vs `pushbytes`, might be to leverage the structure of the bytecblock for some of the key/value properties you're looking for, rather than msgpack.  So, consider:
```
bytecblock "arc-23" 0x32BYTEHEX "arc-XX" 0xSOMEOTHERHASH NNN
where NNN is a 4 byte, bigendian number that says how big the entire bytecblock instruction is.

That would allow clients to find the instruction without diassembling bytecode or understanding messagepack (though they do need to unpack the uvarint encoding of bytecblock). OR if they DO use a disassembler, everything pretty much gets unpacked nicely for them without messagepack or explicit knowledge of uvarint, etc.

Reply:
And yet another option is to insert this special bytecblock after the version byte at the start of the program. That would be safe because jumps are relative, so it would push everything forward.  It would be easier to parse, because it starts at a known location.

The only downside I can think of is that if you disassemble such a modified program, it will have two bytecblocks, which the assembler _might_ complain about.  But if went this way, I would change the assembler if needed.